hybrasyl/Objects/User.cs

```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
-                        {
-                            HealTarget(castable, tar);
-                            spellCast = true;
-                        }
-                    }
-                    if(spellCast) CastingCost(castable);
-                }
-                else
-                {
-                    this.SendSystemMessage("Having no objective in mind, the spell dissipates.");
-                }
-            }
-        }
-
-        public bool canCast(Castable castable)
-        {
-            var costs = castable.CastCosts.CastCost;
-
-            if (double.TryParse(costs.Stat.Hp, out double hpCostPercent))
+            if (UseCastable(castable, targetCreature))
             {
-                if (hpCostPercent != 0)
-                {
-                    uint reduceHp = (uint)Math.Ceiling(this.MaximumHp * hpCostPercent);
-                    if (reduceHp >= this.Hp) return false;
-                }
-            }
-
-            if (double.TryParse(costs.Stat.Mp, out double mpCostPercent))
-            {
-                if (mpCostPercent != 0)
-                {
-                    uint reduceMp = (uint)Math.Ceiling(this.MaximumMp * mpCostPercent);
-                    if (reduceMp > this.Mp) return false;
-                }
-            }
-
-            if (costs.Gold > this.Gold) return false;
-
-
-            if (costs.Items != null)
-            {
-                foreach (var item in costs.Items)
+                Client.Enqueue(new ServerPacketStructures.Cooldown()
                 {
-                    if (item != null)
-                    {
-                        if (!(Inventory.Contains(item.Value, item.Quantity))) return false;
-                    }
-                }
+                    Length = (uint)castable.Cooldown,
+                    Pane = 1,
+                    Slot = slot
+                }.Packet());
             }
-
-            return true;
+            else
+                SendSystemMessage("It didn't work.");
         }
 
-        public void CastingCost(Castable castable)
+        /// <summary>
+        /// Process the casting cost for a castable. If all requirements were not met, return false.
+        /// </summary>
+        /// <param name="castable">The castable that is being cast.</param>
+        /// <returns>True or false depending on success.</returns>
+        public bool ProcessCastingCost(Castable castable)
         {
             var costs = castable.CastCosts.CastCost;
             uint reduceHp = 0;
             uint reduceMp = 0;
             int removeNumItems = 0;
 
-            if (double.TryParse(costs.Stat.Hp, out double hpCostPercent))
-            {
-                if (hpCostPercent != 0)
-                {
-                    reduceHp = (uint)Math.Ceiling(this.MaximumHp * hpCostPercent);
-                }
-            }
+            // HP cost can be either a percentage (0.25) or a fixed amount (50)
+            if (costs.Stat.Hp.Contains('.'))

```

Comment:
May need a Null check here, as Assail does not currently have any stat costs, unless you want to make that a required element, with the Hp / Mp elements being optional.

Reply:
Fixed, also cleaning up a lot of nullrefs elsewhere